tools/src/main/scala/scala/scalanative/compiler/pass/GlobalValueNumbering.scala

```diff
@@ -2,11 +2,283 @@ package scala.scalanative
 package compiler
 package pass
 
-import nir._, Shows._
-import util.sh
+import scala.collection.mutable
 import scala.util.hashing.MurmurHash3
 
-object GlobalValueNumbering {
+import analysis.ClassHierarchy.Top
+import analysis.ClassHierarchyExtractors._
+import analysis.ControlFlow
+import analysis.ControlFlow.Block
+import analysis.DominatorTree
+
+import nir._
+
+class GlobalValueNumbering(implicit top: Top) extends Pass {
+  import GlobalValueNumbering._
+
+  override def preDefn = {
+    case defn: Defn.Define =>
+      val cfg        = ControlFlow.Graph(defn.insts)
+      val domination = DominatorTree.build(cfg)
+
+      val newInsts = performSimpleValueNumbering(cfg, domination)
+
+      Seq(defn.copy(insts = newInsts))
+  }
+
+  private def pred(block: Block) = block.pred.map(_.from)
+  private def succ(block: Block) = block.succ.map(_.to)
```

Comment:
Lets rename `block.pred` into `block.inEdges` and `block.succ` into `block.outEdges` and then put `def pred` and `def succ` as helper methods directly inside `ControlFlow.Block`.